pkg/internal/cnsoperator/cnsfilevolumeclient/cnsfilevolumeclient.go

```diff
@@ -0,0 +1,221 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cnsfilevolumeclient
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/logger"
+	"sigs.k8s.io/vsphere-csi-driver/pkg/internal/cnsoperator/cnsfilevolumeclient/v1alpha1"
+	k8s "sigs.k8s.io/vsphere-csi-driver/pkg/kubernetes"
+)
+
+// FileVolumeClients maintains a client to the API
+// server for operations on CnsFileVolumeClient instance.
+// An in-memory cache is used to read instance data.
+// It also contains a per instance lock to handle
+// concurrent operations.
+type FileVolumeClients struct {
+	client client.Client
+	store  cache.Store
+	// Per volume lock for concurrent access to CnsFileVolumeClient instances.
+	// Keys are strings representing volume handles (or SV-PVC names).
+	// Values are individual sync.Mutex locks that need to be held
+	// to make updates to the CnsFileVolumeClient instance on the API server.
+	fileVolumeClientsLock *sync.Map
+}
+
+// InitFileVolumeClients initializes and returns a variable of type FileVolumeClients.
+func InitFileVolumeClients(ctx context.Context) (*FileVolumeClients, error) {
+	log := logger.GetLogger(ctx)
+	config, err := k8s.GetKubeConfig(ctx)
+	if err != nil {
+		log.Errorf("failed to get kubeconfig. Err: %v", err)
+		return nil, err
+	}
+	k8sclient, err := k8s.NewClientForGroup(ctx, config, v1alpha1.GroupVersion.Group)
+	if err != nil {
+		log.Errorf("failed to create k8s client. Err: %v", err)
+		return nil, err
+	}
+	informer, err := k8s.GetDynamicInformer(ctx, v1alpha1.GroupVersion.Group, v1alpha1.GroupVersion.Version, "cnsfilevolumeclients")
+	if err != nil {
+		log.Errorf("failed to create dynamic informer for cnsfilevolumeclients CRD. Err: %v", err)
+		return nil, err
+	}
+	store := informer.Informer().GetStore()
+	stopCh := make(chan struct{})
+	go informer.Informer().Run(stopCh)
+	if !cache.WaitForCacheSync(stopCh, informer.Informer().HasSynced) {
+		return nil, fmt.Errorf("failed to sync cnsfilevolumeclient cache. Error: %+v", err)
+	}
+	return &FileVolumeClients{
```

Comment:
Is there a need for `InitFileVolumeClients` to return a `FileVolumeClients`? I do not see a need, so I would like to know if I'm missing anything on how FileVolumeClients would be used. Can we use singleton pattern for `FileVolumeClients` ? I think we should use singleton since the `AddClientVMToIPList` and `RemoveClientVMFromIPList` are handling concurrency on a per file volume basis. 

With the current code, I can create any number of FileVolumeClients and invoke various methods on it which defeats the purpose of concurrency handling. I also think we should hide the `FileVolumeClients` struct so that nobody can create this instance outside this package. Any thoughts?

Reply:
So far i've only assumed that CNS Operator will call `InitFileVolumeClients` so i didn't see the need for a singleton it makes sense to add that pattern here too. 

> I also think we should hide the FileVolumeClients struct so that nobody can create this instance outside this package.

My idea has been that CnsFileAccessConfig controller in CNS Operator would have a variable of type `FileVolumeClients`, and it would use that variable to call `AddClientVMToIPList` and `RemoveClientVMFromIPList`. 

If we decide to hide the `FileVolumeClients` struct, would we also want to hide the singleton `InitFileVolumeClients`? Callers just know of `Add...` and `Remove...` and those functions make sure the `fileVolumeClients` object has been initialized before proceeding..

Reply:
>If we decide to hide the FileVolumeClients struct, would we also want to hide the singleton InitFileVolumeClients?

It depends on when you want to allow Add/Remove methods and the behavior of the process when init fails. If you want to fail the process for any errors initializing the FileVolumeClients, then you would need an Init and the main go routine can call the Init method(this initializes the singleton as well). If you want lazy initialization of FileVolumeClients  and cache, and you want to allow Add/Remove only after the cache is synced, then you can init the file volume clients as part of the singleton initialization itself.

I see 2 ways you could implement singleton here:
1. Hide the singleton behind an interface. The callers interact only using the interface. For example see pattern1 usage in https://github.com/SandeepPissay/playground/blob/master/pkg/singleton/singleton.go#L13. This is inline with OOD.
2. Expose only public methods and hide the singleton instance behind them. For example see pattern2 usage in https://github.com/SandeepPissay/playground/blob/master/pkg/singleton/singleton.go#L26

Also see pattern1 and pattern2 packages in https://github.com/SandeepPissay/playground/tree/master/pkg/singleton.

Reply:
@SandeepPissay changed to implement the singleton interface. Can you take a look?